floris/tools/wind_data.py

```diff
@@ -0,0 +1,536 @@
+# Copyright 2024 NREL
+
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not
+# use this file except in compliance with the License. You may obtain a copy of
+# the License at http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+
+# See https://floris.readthedocs.io for documentation
+
+from __future__ import annotations
+
+from abc import abstractmethod
+
+import matplotlib.cm as cm
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+from pandas.api.types import CategoricalDtype
+
+from floris.type_dec import NDArrayFloat
+
+
+class WindDataBase:
+    """
+    Super class that WindRose and TimeSeries inherit from, enforcing the implementation of
+    unpack() on the child classes and providing the general functions unpack_for_reinitialize() and
+    unpack_freq().
+    """
+
+    @abstractmethod
+    def unpack(self):
+        """
+        Placeholder for child classes of WindDataBase, which each need to implement the unpack()
+        method.
+        """
+        raise NotImplementedError("unpack() not implemented on {0}".format(self.__class__.__name__))
+
+    def unpack_for_reinitialize(self):
+        """
+        Return only the variables need for reinitialize
+        """
+        (
+            wind_directions_unpack,
+            wind_speeds_unpack,
+            _,
+            ti_table_unpack,
+            _,
+        ) = self.unpack()
+
+        return wind_directions_unpack, wind_speeds_unpack, ti_table_unpack
+
+    def unpack_freq(self):
+        """Unpack frequency weighting"""
+
+        (
+            _,
+            _,
+            freq_table_unpack,
+            _,
+            _,
+        ) = self.unpack()
+
+        return freq_table_unpack
+
+
+class WindRose(WindDataBase):
+    """
+    In FLORIS v4, the WindRose class is used to drive FLORIS and optimization
+    operations in which the inflow is characterized by the frequency of
+    binned wind speed, wind direction and turbulence intensity values
+
+    Args:
+        wind_directions: NumPy array of wind directions (NDArrayFloat).
+        wind_speeds: NumPy array of wind speeds (NDArrayFloat).
+        freq_table: Frequency table for binned wind direction, wind speed
```

Comment:
The unit tests nicely capture that the frequency has to have the shape `(n wind directions, n wind speeds)`, so that should also be reflected here

Reply:
At least, thats how I understood the unit tests, so let me know if I'm incorrect

Reply:
Done!